crates/red_knot_python_semantic/src/semantic_index/usedef.rs

```diff
@@ -0,0 +1,160 @@
+}
+
+impl<'db> UseDefMap<'db> {
+    pub(crate) fn use_definitions(&self, use_id: ScopedUseId) -> &[Definition<'db>] {
+        &self.all_definitions[self.definitions_by_use[use_id].definitions.clone()]
+    }
+
+    pub(crate) fn use_may_be_unbound(&self, use_id: ScopedUseId) -> bool {
+        self.definitions_by_use[use_id].may_be_unbound
+    }
+
+    pub(crate) fn public_definitions(&self, symbol: ScopedSymbolId) -> &[Definition<'db>] {
+        &self.all_definitions[self.public_definitions[symbol].definitions.clone()]
+    }
+
+    pub(crate) fn public_may_be_unbound(&self, symbol: ScopedSymbolId) -> bool {
+        self.public_definitions[symbol].may_be_unbound
+    }
+}
+
+#[derive(Clone, Debug, PartialEq, Eq)]
+struct Definitions {
+    definitions: Range<usize>,
+    may_be_unbound: bool,
+}
+
+impl Default for Definitions {
+    fn default() -> Self {
+        Self {
+            definitions: 0..0,
+            may_be_unbound: true,
+        }
+    }
+}
+
+#[derive(Debug)]
+pub(super) struct FlowSnapshot {
+    definitions_by_symbol: IndexVec<ScopedSymbolId, Definitions>,
+}
+
+pub(super) struct UseDefMapBuilder<'db> {
+    all_definitions: Vec<Definition<'db>>,
+
+    definitions_by_use: IndexVec<ScopedUseId, Definitions>,
+
+    // builder state: currently visible definitions for each symbol
+    definitions_by_symbol: IndexVec<ScopedSymbolId, Definitions>,
+}
+
+impl<'db> UseDefMapBuilder<'db> {
+    pub(super) fn new() -> Self {
+        Self {
+            all_definitions: Vec::new(),
+            definitions_by_use: IndexVec::new(),
+            definitions_by_symbol: IndexVec::new(),
+        }
+    }
+
+    pub(super) fn add_symbol(&mut self, symbol: ScopedSymbolId) {
+        let new_symbol = self.definitions_by_symbol.push(Definitions::default());
+        debug_assert_eq!(symbol, new_symbol);
+    }
+
+    pub(super) fn record_def(&mut self, symbol: ScopedSymbolId, definition: Definition<'db>) {
+        let def_idx = self.all_definitions.len();
+        self.all_definitions.push(definition);
+        self.definitions_by_symbol[symbol] = Definitions {
+            #[allow(clippy::range_plus_one)]
+            definitions: def_idx..(def_idx + 1),
+            may_be_unbound: false,
+        };
+    }
+
+    pub(super) fn record_use(&mut self, symbol: ScopedSymbolId, use_id: ScopedUseId) {
+        let new_use = self
+            .definitions_by_use
+            .push(self.definitions_by_symbol[symbol].clone());
+        debug_assert_eq!(use_id, new_use);
+    }
+
+    pub(super) fn snapshot(&self) -> FlowSnapshot {
+        FlowSnapshot {
+            definitions_by_symbol: self.definitions_by_symbol.clone(),
+        }
+    }
+
+    pub(super) fn set(&mut self, state: &FlowSnapshot) {
+        let num_symbols = self.definitions_by_symbol.len();
+        self.definitions_by_symbol = state.definitions_by_symbol.clone();
+        self.definitions_by_symbol
+            .resize(num_symbols, Definitions::default());
```

Comment:
When is it possible that `num_symbols` > self.definitions_by_symbol.len()`? Is `Definitions::default` really the right default in that case. 

You can use  `truncate` if it is guaranteed that `num_symbols < self.definitions_by_symbol.len` (and it would probably be worth adding an assertion)

Reply:
(Feel free to either read and reply to comments here, or just wait for the follow up PR with module docs, whatever you prefer. I reply here first either way because it helps me to sort out what I need to write in the module docs.)

The purpose of this method is to reset the builder state to some previous snapshot. For instance, when we have `if a: body_code else: else_code`, after we visit `body_code` we have to reset the visible-definitions state to what it was before the `if`, before we visit `else_code`, since either `body_code` or `else_code` can run, but not both, so definitions in `body_code` can never be visible to `else_code`.

But we also have to make sure that every symbol we have seen (even in `body_code`) is included in `symbols_by_definition`, otherwise our symbol IDs in the IndexVec will get out of sync: we might try to record some new symbol in `else_code` with the wrong symbol ID, because we are missing the ones from `body_code`.

Like say we have this:

```
if flag:
    x = 1
else:
    y = 2
```

After visiting `x = 1` we need to visit the else clause with no visible definition for `x`. We reset the state to our snapshot from before the `if`, which doesn't even include `x` in `definitions_by_symbol` yet. But then when we visit `y = 2` we will insert it to the wrong slot in `definitions_by_symbol`, because we are missing `x` -- the symbol IDs don't line up correctly.

Because we only add symbols as we visit code, never remove them, the current `num_symbols` can never be less than the number of symbols in the earlier snapshot we are resetting to, only greater (e.g. in the above example `num_symbols` is 1 -- it includes `x` -- but the snapshot we are resetting to has zero symbols). So the only case that can occur here is that we resize the snapshot from zero to one, and add a "no definitions" for `x`. So `Definitions::default()` is correct here, because it represents "no visible definitions" (an empty range, with `may_be_unbound: true`).

In the follow-up PR, in addition to the module doc comment, I'll add assertions and comments throughout this code.

Reply:
Hopefully clearer with the docs in https://github.com/astral-sh/ruff/pull/12357